Remove logrus

rosstimothy opened the pull request:
The only remaining use of logrus is from integrations, which is unfortunately imported by teleport.e, and prevents logrus from being moved to an indirect dependency. The logrus formatter and initialization of the logrus logger will remain in place until integrations is using slog. To prevent any accidental inclusions of logrus within the teleport module the depguard rules have been updated to prohibit importing logrus. The rules also include prohibit a few common log packages that tools like gopls might automatically import.


```bash
$ go mod why github.com/sirupsen/logrus
# github.com/sirupsen/logrus
github.com/gravitational/teleport/integrations/access/common/auth
github.com/sirupsen/logrus
```

Includes https://github.com/gravitational/teleport.e/pull/5807
Depends on https://github.com/gravitational/teleport/pull/50805

rosstimothy committed:
Remove all direct logrus usage from teleport module

The only remaining use of logrus is from integrations, which is
unfortunately imported by teleport.e, and prevents logrus from
being moved to an indirect dependency. The logrus formatter and
initialization of the logrus logger will remain in place until
integrations is using slog. To prevent any accidental inclusions
of logrus within the teleport module the depguard rules have been
updated to prohibit importing logrus. The rules also include
prohibit a few common log packages that tools like gopls might
automatically import.

strideynet commented on tool/tbot/spiffe.go:
```diff
@@ -21,22 +21,46 @@ package main
-	"github.com/gravitational/teleport/lib/utils"
 )
 
+// TODO(tross/noah): Remove once go-spiff has a slog<->workloadapi.Logger adapter.
```
I have a ticket w/ go-spiffe to add slog support - https://github.com/spiffe/go-spiffe/issues/281

What you've done here is good for now!

codingllama commented on .golangci.yml:
```diff
@@ -117,16 +113,29 @@ linters-settings:
+          # been converted to slog yet. Once they use slog, remove this exception.
+          - '!**/integrations/**'
+          # The log package still contains the logrus formatter consumed by the integrations.
+          # Remove this exeception when said formatter is deleted.
```
```suggestion
          # Remove this exception when said formatter is deleted.
```

codingllama commented on e:
Friendly reminder that e/ is already past this commit, in case you want to pull the latest one.

codingllama commented on lib/utils/log/slog.go:
```diff
@@ -34,8 +37,56 @@ const (
 	// TraceLevelText is the text representation of Trace verbosity.
 	TraceLevelText = "TRACE"
+
+	noColor = -1
```
Is it worth keeping the added code in a file other than slog.go?

Are there parts of it we expect to delete alongside the last remains of logrus? Those could be in a separate file too (that could be deleted in its entirety).

codingllama replied:
Sounds good.

codingllama commented on lib/utils/log/slog.go:
```diff
@@ -68,6 +119,45 @@ func addTracingContextToRecord(ctx context.Context, r *slog.Record) {
+	TimestampField: {},
+}
+
+func ValidateFields(formatInput []string) (result []string, err error) {
```
Add a godoc?

codingllama commented on tool/tbot/spiffe.go:
```diff
@@ -21,22 +21,48 @@ package main
+}
+
+func (l logger) Debugf(format string, args ...interface{}) {
+	//nolint:sloglint // msg cannot be constant
+	l.l.DebugContext(context.Background(), fmt.Sprintf(format, args...))
```
Maybe:

```go
	l.log(slog.LevelDebug, format, args...)
```

and implement the l.log function so it centralized the nolint / fmt.Sprintf logic?

codingllama commented on tool/tbot/spiffe.go:
```diff
@@ -21,22 +21,48 @@ package main
 		// TODO(noah): Upstream PR to add slog<->workloadapi.Logger adapter.
-		workloadapi.WithLogger(utils.NewLogger()),
+		// https://github.com/spiffe/go-spiffe/issues/281
+		workloadapi.WithLogger(logger{l: slog.Default()}),
```
Maybe use a specialized component for this logger, or some other distinctive With key/value?

codingllama replied:
Happy to keep it as-is if you'd rather avoid other changes here.

strideynet replied:
I'm happy with this for now. I'll revisit all of tbot daemon logs at somepoint

codingllama commented on tool/teleport/common/teleport.go:
```diff
@@ -1017,7 +1017,7 @@ func dumpConfigFile(outputURI, contents, comment string) (string, error) {
 	// when 'teleport scp' is executed, it cannot write logs to stderr (because
 	// they're automatically replayed by the scp client)
-	utils.SwitchLoggingToSyslog()
+	slog.SetDefault(slog.New(logutils.DiscardHandler{}))
```
Is it a problem that these aren't equivalent anymore?

codingllama replied:
No strong opinions on my part but attempting syslog first does seem better from a behavioral-compatibility point of view.

codingllama replied:
Sounds OK.

codingllama reviewed (commented):
Nice! We are almost 100% logrus-free!

rosstimothy committed:
Refactor logger initialization

Consolidates configuring of global loggers to a single function.
This is mainly to facilitate configuring the logger for
teleport scp, but will also allow us to remove the copy of logger
initialization that currently exists in integrations/lib/logger.

rosstimothy committed:
fix: document ValidateFields

bl-nero commented on tool/teleport/common/teleport.go:
```diff
@@ -1014,10 +1014,22 @@ func dumpConfigFile(outputURI, contents, comment string) (string, error) {
+	})
+	if err != nil {
+		// If something went wrong, discard all logs and continue command execution.
+		slog.SetDefault(slog.New(logutils.DiscardHandler{}))
```
That's different from the current implementation that (1) printed the error in such case, and (2) rerouted logs to stderr. Why did you change it?

codingllama commented on lib/config/configuration.go:
```diff
@@ -789,79 +781,22 @@ func applyAuthOrProxyAddress(fc *FileConfig, cfg *servicecfg.Config) error {
-		cfg.Console = io.Discard // disable console printing
-	case "stdout", "out", "1":
-		w = os.Stdout
+	case "stderr", "error", "2", "stdout", "out", "1":
```
Maybe keep the cases for `""` and `teleport.Syslog` in the meantime? Is the case for `teleport.Syslog` obsolete too?

codingllama commented on lib/utils/log/log.go:
```diff
@@ -0,0 +1,128 @@
+	// Severity defines how verbose the log will be. Possible values are "error", "info", "warn"
+	Severity string `yaml:"severity,omitempty"`
+	// Format defines the output format. Possible values are 'text' and 'json'.
+	Format string `yaml:"output,omitempty"`
```
Yaml tag conflicts with Output.

codingllama commented on lib/utils/log/log.go:
```diff
@@ -0,0 +1,128 @@
+			return slog.Default(), level, nil
+		}
+	default:
+		// Assume this is a file path.
```
```suggestion
		// Assume loggerConfig.Output is a file path.
```

codingllama commented on integrations/lib/logger/logger.go:
```diff
@@ -108,8 +99,16 @@ func Setup(conf Config) error {
-// TODO: this code is adapted from `config.applyLogConfig`, we'll want to deduplicate the logic next time we refactor the logging setup
+// TODO(tross): Defer logging initialization to logutils.Initialize and use the
+// global slog loggers once integrations has been updated to use slog.
 func (conf Config) NewSLogLogger() (*slog.Logger, error) {
```
pre-existing nit: s/SLog/Slog, ie `NewSlogLogger`.

codingllama commented on lib/utils/log/slog.go:
```diff
@@ -68,6 +119,47 @@ func addTracingContextToRecord(ctx context.Context, r *slog.Record) {
+
+// ValidateFields ensures the provided fields map to the allowed fields. An error
+// is returned if any of the fields are invalid.
+func ValidateFields(formatInput []string) (result []string, err error) {
```
Could this be made private and made a responsibility of Initialize? Perhaps after the duplicated code on integrations/ is removed?

rosstimothy committed:
fix: remove copied yaml tags

rosstimothy committed:
fix: update file path comment
